crypto/s2n_aead_cipher_chacha20_poly1305.c

```diff
@@ -157,6 +156,106 @@ static int s2n_aead_chacha20_poly1305_destroy_key(struct s2n_session_key *key)
     return 0;
 }
 
+#elif defined(S2N_CHACHA20_POLY1305_AVAILABLE_BSSL_AWSLC) /* BoringSSL and AWS-LC implementation */
+
+static int s2n_aead_chacha20_poly1305_encrypt(struct s2n_session_key *key, struct s2n_blob *iv, struct s2n_blob *aad, struct s2n_blob *in, struct s2n_blob *out)
+{
+    gte_check(in->size, S2N_TLS_CHACHA20_POLY1305_TAG_LEN);
+    gte_check(out->size, in->size);
+    eq_check(iv->size, S2N_TLS_CHACHA20_POLY1305_IV_LEN);
+
+    /* Adjust input length to account for the Tag length */
+    size_t in_len = in->size - S2N_TLS_CHACHA20_POLY1305_TAG_LEN;
+    size_t out_len = 0;
+
+    GUARD_OSSL(EVP_AEAD_CTX_seal(key->evp_aead_ctx, out->data, &out_len, out->size, iv->data, iv->size, in->data, in_len, aad->data, aad->size), S2N_ERR_ENCRYPT);
+
+    S2N_ERROR_IF((in_len + S2N_TLS_CHACHA20_POLY1305_TAG_LEN) != out_len, S2N_ERR_ENCRYPT);
+
+    return 0;
+}
+
+static int s2n_aead_chacha20_poly1305_decrypt(struct s2n_session_key *key, struct s2n_blob *iv, struct s2n_blob *aad, struct s2n_blob *in, struct s2n_blob *out)
+{
+    gte_check(in->size, S2N_TLS_CHACHA20_POLY1305_TAG_LEN);
+    gte_check(out->size, in->size);
+    eq_check(iv->size, S2N_TLS_CHACHA20_POLY1305_IV_LEN);
+
+    size_t out_len = 0;
+
+    GUARD_OSSL(EVP_AEAD_CTX_open(key->evp_aead_ctx, out->data, &out_len, out->size, iv->data, iv->size, in->data, in->size, aad->data, aad->size), S2N_ERR_DECRYPT);
+
+    S2N_ERROR_IF((in->size - S2N_TLS_CHACHA20_POLY1305_TAG_LEN) != out_len, S2N_ERR_ENCRYPT);
```

Comment:
Should you also set `out->size` to `out_len`? Or change line 181 to `eq_check(out->size, in->size - S2N_TLS_CHACHA20_POLY1305_TAG_LEN);`?

Is it possible s2n's accounting could be wrong and pass in a blob too big? AWS-LC could decrypt a message that is smaller than the out blob and s2n would attempt to use the data at the end of out blob.

Same comment for the other 3 encrypt/decrypt methods.

Reply:
Good point, the reason the check in line 181 is always true is because decryption is performed in-place. `out->size` should be at least the input size minus the tag length.

Encryption is also performed in-place, but here `out->size` should be at least the input size plus the tag length.

Reply:
`in-size` includes the size of the data and the MAC tag. So changes the check of `out->size` in the encryption function to:
```
gte_check(out->size, in->size + S2N_TLS_CHACHA20_POLY1305_TAG_LEN);
```
would not be correct.

I adjusted the check of `out->size` in the decryption function to not include the MAC tag length, since the output buffer should not be expected to carry this.

Reply:
I agreed with @andrewhop . While these methods are being used in place the API does not require this. Please ensure that this method works properly when not-used in-place. (Added `out->size = out_len` as suggested would probably fix this.)

Reply:
I am adding support for ChaCha20-Poly1305. Going through these functions to audit whether they support in-place / not in-place buffers and making changes to how they can be called is out-of-scope. If we want to add this functionality we would also need to consider extending the set unit tests and consider functions in other parts of the S2N source code that would need to adopt these changes. This is a logically different change.

That being said, there is nothing in this particular function that prevents using disjoint buffers. Encrypt functions generally assumes the `in` blob has a particular encoding that is not compatible with this e.g. assuming:
```C
    /* Adjust input length to account for the Tag length */
    size_t in_len = in->size - S2N_TLS_CHACHA20_POLY1305_TAG_LEN;
```

> Added out->size = out_len as suggested would probably fix this

Could you expand on what you mean? This check will break the current code.

Reply:
What would happen if I called this method where `in->size == 256` and `out->size == 512`?

1. Checks on 180-182 would pass
2. Line 186 would succeed and would set `out_len = 240`
3. Check on line 188 would pass
4. Method returns successfully with `out->size == 512` which doesn't match the amount of data decrypted.

Is this a problem? (I'm still learning the s2n code-base so may be misunderstanding some of the requirements we're trying to manage.)